---
title: "proj_test"
author: ""
date: '202X-XX-XX'
output: html_document
---

```{r setup, include=FALSE}
library(reticulate)
library(MetaBAnalysis)
knitr::opts_chunk$set(echo = TRUE)
knitr::knit_engines$set(python = reticulate::eng_python)
```

### Filter the data and run dada2:
```{r}
# Set project name:
projId <- "XXX_202X"

# Collect sample paths and names. Duplicate line once per primer:
primer1 <- CollectData2("../Filtered_data", prefix = "mifish")
# primer2 <- CollectData2("../Filtered_data", prefix = "v16s")

# Filter reads and generate out object for later use:
out <- FiltTrimWrap(primer1)
saveRDS(out, file = "out.rds")
# out2 <- FiltTrimWrap(primer2)
# saveRDS(out2, file = "out2.rds")

# Dereplicate and merge pairs with dada2:
dada2Counts <- DadaAnalysis2(primer1, justConcatenate = FALSE, minOverlap = 5)
# dada2Counts2 <- DadaAnalysis2(primer2, justConcatenate = FALSE, minOverlap = 5)
```

### Convert and modify matrix "dada2Counts" to DGEList "yAll":
```{r}
yAll <- MakeDGEList2(dada2Counts, primer1)
saveRDS(yAll, file = "yAll.rds")
# yAll2 <- MakeDGEList2(dada2Counts2, primer2)
# saveRDS(yAll2, file = "yAll2.rds")
```

### Export fasta-file:
```{r}
ExportFasta(yAll, paste0("y_", projId, primer1$Prefix, ".fa"), minLength = 30, maxLength = 1000)
# ExportFasta(yAll2, paste0("y_", projId, primer2$Prefix, ".fa"), minLength = 30, maxLength = 1000)
```

### BLAST-tools (Experimental):
```{r}
# blastOut <- MultiBlaster(fastaFile = "y_test.fa")
# BlastResWriter(blastOut, "y_test.out")
```

### Parse BLAST-results:
```{r}
### The following code assumes that the folder "../../Suppl_data" exists and contains the files "compact_names.csv" and
# "compact_nodes.csv. If it does not, create the folder in that location and perform the following:
# 1. $ cd Suppl_data
# 2. $ wget https://ftp.ncbi.nih.gov/pub/taxonomy/taxdmp.zip
# 3. $ unzip taxdmp.zip
# 4. $ awk -F\t '{gsub(/[^[:alnum:]]/," ", $3); if ($7 == "scientific name") {print $1 "\t" $3}}' names.dmp > compact_names.csv
# 5. $ awk -F\t '{gsub(/[^[:alnum:]]/," ", $3); print $1 "\t" $3 "\t" $5}' nodes.dmp > compact_nodes.csv
# 6. $ rm *.dmp *.prt *.txt *.zip

data(compactNameDump, package = "MetaBAnalysis")
data(compactNodeDump, package = "MetaBAnalysis")

blastResY <- MetaBAnalysis::BlastParse(yAll, paste0("y_", projId, primer1$Prefix, ".out"), threshold = 90)
saveRDS(blastResY, file = "blastResY.rds")
# blastResY2 <- MetaBAnalysis::BlastParse(yAll2, paste0("y_", projId, primer2$Prefix, ".out"), threshold = 90)
# saveRDS(blastResY2, file = "blastResY2.rds")
```

### Create Summaryfiles from the BLAST-results:
```{r}
allResults <- MetaBAnalysis::SumRes(blastRes = blastResY, counts = yAll$counts, taxGroup = "Fish") # Summarize a list of result objects
# allResults2 <- MetaBAnalysis::SumRes(blastRes = blastResY2, counts = yAll2$counts, taxGroup = "Fish") # Summarize a list of result objects
```

# Rstudio functions:
```{r}
# These functions are intended to aid the curation step by allowing the user to
# quickly write the sequences related to a species name to a file (QSeqGetter
# and QSSplit) or blast the top three sequences related to a species and View the
# results in Rstudio (QSBLAST).
library(MetaBAnalysis)

# Write sequences to file
`1QSeqGetter` <- function(speciesName, blastResults = blastResY) {
  species_seq <- paste0(">",
                        blastResults[grepl(speciesName, blastResults$species),1],
                        "\n",
                        blastResults[grepl(speciesName, blastResults$species),2])
  writeLines(species_seq, "current_sequence.txt")
  file.edit("current_sequence.txt")
}

`2QSeqGetter` <- function(x) {`1QSeqGetter`(x, blastResY2)} # wrapper for second primer

# Write sequences to file and split sequences that contains 'NNNNNNNNNN'
`1QSSplit` <- function(speciesName, blastResults = blastResY) {
  seqName <- blastResults[grepl(speciesName, blastResults$species),1]
  seqFull <- blastResults[grepl(speciesName, blastResults$species),2]
  seq1 <- lapply(strsplit(seqFull, "NNNNNNNNNN"), "[[", 1)
  seq2 <- lapply(strsplit(seqFull, "NNNNNNNNNN"), "[[", 2)
  species_seq <- paste0(">", seqName,
                        "\n", seqFull, "\n",
                        ">", seqName, ".1",
                        "\n", seq1, "\n",
                        ">", seqName, ".2",
                        "\n", seq2)
  writeLines(species_seq, "current_sequence.txt")
  file.edit("current_sequence.txt")
}

`2QSSplit` <- function(x) {`1QSSplit`(x, blastResY2)} # wrapper for second primer

# Blast the top three sequences of a species # Does not work
QSBLAST <- function(searchName, seqNumber = 3, blastResults = blastResY) {
  QSeqGetter(searchName, blastResults)
  blastOut <- MultiBlaster("current_sequence.txt",
                           seqNumber,
                           resultNumber = 10,
                           viewChoice = "YES")
  View(blastOut)
}
```

### Curate the results (first primer):
```{r}
library(MetaBAnalysis)
library(rvest)

{
  ### Create object:
  groupNotes <- MakeNotes(allResults$TargetGroup)
  
  ### Filter:
  groupNotes <- TrimCutoffs(total = 0.1, within = 0.1, minimum = 0, groupNotes) # trim species based on cutoffs
  
  ### Curate:
  groupNotes <- CurateHit(ref = 'hitX', note = '', absorb = '', latin = '', common = '', rmHit = FALSE, groupNotes)
  
  ### Create outputs:
  groupClean <- groupNotes[, -c(1, 2, 5)]
  columnNames <- colnames(groupClean)
  columnNames[-c(1, 2)] <- lapply(strsplit(columnNames[-c(1, 2)], "_"), `[`, 2) # remove first underscore
  # columnNames <- gsub("pattern", "replacement", columnNames)
  # columnNames[grepl("old sample name", columnNames)] <- "new sample name"
  names(groupClean) <- columnNames
  allResults$Clean <- groupClean
}

saveRDS(allResults, "allResults.rds") # save the clean object to be used in the report
```

### Curate the results (second primer, if relevant):
```{r}
library(MetaBAnalysis)
library(rvest)

{
  ### Create object:
  groupNotes2 <- MakeNotes(allResults2$TargetGroup)
  
  ### Filter:
  groupNotes2 <- TrimCutoffs(total = 0.1, within = 0.1, minimum = 0, groupNotes2) # trim species based on cutoffs
  
  ### Curate:
  groupNotes2 <- CurateHit(ref = 'hitX', note = '', absorb = '', latin = '', common = '', rmHit = FALSE, groupNotes2)
  
  ### Create outputs:
  groupClean2 <- groupNotes2[, -c(1, 2, 5)]
  columnNames2 <- colnames(groupClean2)
  columnNames2[-c(1, 2)] <- lapply(strsplit(columnNames2[-c(1, 2)], "_"), `[`, 2) # remove first underscore
  # columnNames2 <- gsub("pattern", "replacement", columnNames2)
  # columnNames2[grepl("old sample name", columnNames2)] <- "new sample name"
  names(groupClean2) <- columnNames2
  allResults2$Clean <- groupClean2
}

saveRDS(allResults2, "allResults2.rds") # save the clean object to be used in the report
```

### Generate information about the sequences for use in report:
```{r}
library(rvest)

rawSeqsTable <- MetaBAnalysis::getRawSeqs() # add searchPattern = "_fastqc.html" if single-end data

rawSeqs <- rawSeqsTable$nreads

sum(rawSeqs)
min(rawSeqs)
max(rawSeqs)

primerSeqs <- sum(out[, 1])
primerSeqs / sum(rawSeqs) # ratio of sequences left after primer trimming and filtration

groupSeqs <- sum(groupClean[, -c(1, 2)])
groupSeqs / primerSeqs # ratio of filtered sequences belonging to target group

sampleMetaData <- list(rawSeqs = rawSeqs, primerSeqs = primerSeqs, groupSeqs = groupSeqs)
saveRDS(sampleMetaData, "sampleMetaData.rds")

sum(groupClean[1, -c(1, 2)]) / sum(groupClean[, -c(1, 2)]) # see what percentage of counts the top four species accounts for
sum(groupClean2[1, -c(1, 2)]) / sum(groupClean2[, -c(1, 2)]) # see what percentage of counts the top four species accounts for

if("allResults2" %in% ls()) { # if second primer
  sampleMetaData$primerSeqs2 <- sum(out2[, 1]) # add n seqs with primer for second primer
  sampleMetaData$groupSeqs2 <- sum(groupClean2[, -c(1, 2)])
  saveRDS(sampleMetaData, "sampleMetaData.rds")
}
```

### Generate excel-sheet with counts from "groupClean":
```{r}
library(writexl)
write_xlsx(groupClean, paste0("projekt_", projId,"_sekvensantal_", primer1$Prefix, ".xlsx"))
# write_xlsx(groupClean2, paste0("projekt_", projId,"_sekvensantal_", primer2$Prefix, ".xlsx"))
```